[PROTOCOL RFC] Full Void Type Support (#7073)
delta-io/delta: protocol_rfcs/void-type.md

# Void Type
**Associated Github issue for discussions: https://github.com/delta-io/delta/issues/7072**

> The `voidType` reader/writer table feature adds support for using the `void` data type (also known as `NullType` in Spark, `UnknownType` in Iceberg, and `UNKNOWN` in Parquet) anywhere in a Delta table schema.

Reviewer: The name `voidType` suggests that it's required for all `VOID` columns in the table, but this is actually not case. Maybe `storedVoidType` or `materializedVoidType`?

Author: `materializedVoidType` sounds good. Will update to that.

`void` is a data type with a single possible value: `NULL`. A column ends up with this type when the writer has no information about its actual type, typically because every value observed so far has been `NULL` (for example, `CREATE TABLE t AS SELECT NULL AS a`, or schema evolution that adds a column containing only `NULL`s).

Today, `void` columns are represented by omitting them from data files and reconstructing them as all-`NULL` columns on read (the missing columns mechanism). That representation cannot encode four schema shapes - a table whose columns are all `void`, a `struct` whose fields are all `void`, a `void` nested in an `array`, and a `void` nested in a `map` - because in each case omitting the `void` column(s) would leave the enclosing `struct`, `array`, or `map` (or the table itself) with nothing written to a data file, and therefore nowhere to record whether the enclosing value is `NULL`, empty, or how long it is. Writers must reject writing data in those cases.

> The `voidType` table feature lifts these restrictions by storing the `void` columns those shapes require - the **structural** `void` columns - using the Parquet [`UNKNOWN` logical type](https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#unknown-always-null): a column whose values are always `NULL` but which is physically present in the data file, so it can carry the information about its enclosing complex value. Older clients cannot read `UNKNOWN` columns, so the representation is gated behind this reader/writer feature.

Reviewer: Should we also store the `void` columns in the data files using the `UNKNOWN` type in the other cases?

Author: Discussing in https://github.com/delta-io/delta/pull/7073#discussion_r3491430529

- a `void` directly inside an `array` or `map` (at any nesting level): the `void` element or value is structural and must be stored as `UNKNOWN`.
- a `struct` whose fields are all `void`, or a table whose columns are all `void`: the writer chooses which `void` column is structural and must store it as `UNKNOWN`.

A `void` column in any other position is never structural: it can be omitted, and does not require the feature. A schema that contains one of the shapes above is said to **require** the `voidType` feature.

### Writer Requirements for Void Type

When Void Type is supported (when the `writerFeatures` field of a table's `protocol` action contains `voidType`), writers:
- must write the table's structural `void` columns to data files (see [Structural void columns](#structural-void-columns)).

> - should omit any non-structural `void` column.

Reviewer: ```suggestion
- should omit any non-structural `void` column from data files.
```

> - should omit any non-structural `void` column.

Reviewer: We should have a debate about this. By always writing the `void` columns we can reduce the risk of a buggy implementation. The downside is additional storage costs,  but these should be minimal, as the null values can be compressed to just a couple of bytes using RLE encoding.

Author: Parquet does not store null values, so no encoding needed for the data itself. It only stores the definition / repetition levels for UNKNOWN, so the cost is minimal. But:
- Minimal cost is still a cost and can appear as a regression compared to when the feature is off if the table has huge number of voids.
- Removing the table feature becomes harder when we use `UNKNOWN` for all voids, as we need to make sure the table does not have any `UNKNOWN` when we drop the feature. If we only use `UNKNOWN` for structural columns, then the user only needs to drop/widen those void columns for dropping the feature.

### Reader Requirements for Void Type

When Void Type is supported (when the `readerFeatures` field of a table's `protocol` action contains `voidType`), readers:
- must allow a `void` data type anywhere in a Delta table schema.
- must return only `NULL` values for a `void` column regardless of how it is represented.
- must, within a single scan, correctly combine data files that represent the same column differently - omitted, written as an all-`NULL` column, or (after a type change) written with a concrete type - into the requested read schema.

### Removing the Void Type feature

Because clients that do not support the feature cannot read `UNKNOWN` columns, removing `voidType` requires that the table no longer depend on the `UNKNOWN` representation. In the version that removes `voidType` from the `writerFeatures` and `readerFeatures` fields of the table's `protocol` action, writers:
- must ensure that the table schema does not require the feature - it must contain none of the shapes in [Structural void columns](#structural-void-columns).

> - must ensure that no data file reachable by the table (including via time travel within the retained history) contains a column stored as `UNKNOWN`. This may require rewriting existing data files so that every `void` column is represented by omission.

Reviewer: ```suggestion
- must ensure that no data file reachable by the table (including via time travel within the retained history) contains a column stored as `UNKNOWN`. This may require rewriting existing data files and/or truncating the table history so that every `void` column is represented by omission in order to be able to remove the `voidType` feature.
```

> After the feature is removed, the table reverts to representing `void` columns only by omission, and the shapes that require the feature are again rejected when writing data.

Reviewer: ```suggestion
After the feature is removed, the table reverts to representing `void` columns only by omission, and the shapes that require the feature must again be rejected when writing data or altering the table schema.
```

Author: Note that altering the table schema is okay without writing data when we can't support the schema. Metadata-only operations can work with the bad schema, only actual data writes fail.

I'll reword the sentence to use `must`.

# Void Type
**Associated Github issue for discussions: https://github.com/delta-io/delta/issues/7072**

> This protocol change adds support for using the `void` data type (also known as `NullType` in Spark, `UnknownType` in Iceberg, and `UNKNOWN` in Parquet) anywhere in a Delta table schema, via a new reader/writer table feature, `voidType`.

Reviewer 2: ```suggestion
The `voidType` reader/writer table feature adds support for using the `void` data type (also known as `NullType` in Spark, `UnknownType` in Iceberg, and `UNKNOWN` in Parquet) anywhere in a Delta table schema.
```

`void` is a data type with a single possible value: `NULL`. A column ends up with this type when the writer has no information about its actual type, typically because every value observed so far has been `NULL` (for example, `CREATE TABLE t AS SELECT NULL AS a`, or schema evolution that adds a column containing only `NULL`s).

> Today, `void` columns are represented by omitting them from data files and reconstructing them as all-`NULL` columns on read (the missing columns mechanism). That representation cannot encode four schema shapes - a table whose columns are all `void`, a `struct` whose fields are all `void`, a `void` nested in an `array`, and a `void` nested in a `map` - because in each case omitting the `void` column(s) would leave the enclosing `struct`, `array`, or `map` (or the table itself) with nothing written to a data file, and therefore nowhere to record whether the enclosing value is `NULL`, empty, or how long it is. Writers must reject writing data in those cases.

Reviewer 2: That's true only for some engines like Spark. I think it's officially undefined since it is not supported.

Author: It's becoming official in https://github.com/delta-io/delta/pull/6966. I built this RFC assuming that protocol clarification makes it in.

Before that change, the protocol basically said tables can have `void`, behavior is undefined but it's recommended to drop it upon reads.

--------

> ***New section after the [Variant Shredding](/PROTOCOL.md#variant-shredding) section***

# Void Type

`void` is a primitive data type (see [Primitive Types](/PROTOCOL.md#primitive-types)) with a single possible value, `NULL`. A `void` column can be represented in a data file in one of two ways:

- **Omitted** - the column is not written to the data file and is reconstructed on read as an all-`NULL` column, following the [rule](/PROTOCOL.md#consistency-between-table-metadata-and-data-files) that a column present in the table schema but absent from a data file is read as `NULL`.
- **Stored as `UNKNOWN`** - the column is written using the Parquet [`UNKNOWN` logical type](https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#unknown-always-null). Its values are always `NULL`, but unlike an omitted column it is physically present in the data file. This representation requires the `voidType` table feature.

A `void` column may be changed to any other data type through supported schema-evolution operations; this does not require the [Type Widening](/PROTOCOL.md#type-widening) table feature, even when the `void` column is stored as `UNKNOWN`.

> ## Void columns without the table feature

Reviewer 2: I'm not sure we can enforce anything on the case without a table feature. A legacy writer/reader does not have knowledge of this new proposal and cannot retroactively ban certain operations.

It seems to me, if someone wants the void type to behave as expected, they must have the table feature and from there it's up to the engine whether they want to omit or materialize the void type column.

Author: The situation is not ideal. `void` never officially made it into the protocol, but got accidentally introduced to tables by the Spark connector when Spark got NullType support. Then there were various revisions to the Protocol to make sure `void` is mentioned there, but it was all vague because even the Spark connector did not handle it properly causing query failures. After https://github.com/delta-io/delta/pull/6966, the behavior will be defined, and both Spark connector and kernel-rs follows that version of the Protocol.

I understand external clients may now become protocol-incompliant, but if they somehow managed to read what is written by Spark (which I think is the reference implementation) previously, and if they wrote something that Spark could read before, then they should still be protocol-compliant. In any case, this comment is more for https://github.com/delta-io/delta/pull/6966 than this PR.

When the `voidType` feature is not supported, `void` columns can only be **omitted**. Because a `void` column is never written to a data file, writers must reject **writing data** to a table whose schema contains any of the following shapes, in which omitting the `void` column(s) would leave nowhere to record the nullability or length of an enclosing value:

> - a `void` type directly inside an `array` or `map` at any nesting level;

Reviewer 3: For `map`, shall we specify that it is allowed only for the value? I recall that we don't allow VOID keys anyway, right?

Also, by 'directly,' we mean arrays like `ARRAY<VOID>`. But with this feature, are we also going to unblock `void` indirectly inside an `array` or `map`, right? Like `ARRAY<STRUCT<INT, VOID>>`?

Author: > For `map`, shall we specify that it is allowed only for the value? I recall that we don't allow VOID keys anyway, right?

That is Spark limitation, and I don't think it belongs in Delta Protocol. FWIW The Protocol does not say anything about nullability of map keys in general.

> Also, by 'directly,' we mean arrays like `ARRAY<VOID>`. But with this feature, are we also going to unblock `void` indirectly inside an `array` or `map`, right? Like `ARRAY<STRUCT<INT, VOID>>`? 

Indirect voids are already unblocked, it's just an implementation detail that Spark connector blocks this, but otherwise there's no reason for blocking. This feature does not affect those voids.

- a `struct` (at any nesting level) whose fields are all `void`; or
- a table whose columns are all `void`.

> These restrictions are stated in terms of the **table schema**, not the schema of any individual data file. A table with such a schema can still be created, altered through metadata-only operations, and read. It can be made writable by evolving its schema - for example, by changing a `void` column to another type - or by enabling the `voidType` feature.

Reviewer 2: If the column is omitted, then the schema of the individual data files are all the same, right?

Author: There can be differences due to schema evolution/type widening. Voids will always be missing, but other columns can differ.

2. A client that supports `voidType` is also guaranteed to correctly read and write structural `void` columns (those stored as `UNKNOWN`). Enabling the feature is **required** to write data for a schema that needs the `UNKNOWN` representation, because clients that do not support the feature cannot read `UNKNOWN` columns.

### Structural void columns

A `void` column is **structural** when it must be stored as `UNKNOWN` rather than omitted, because omitting it would leave an enclosing `struct`, `array`, or `map` (or the table) with nothing written to the data file. This arises only in the schema shapes that the missing columns mechanism cannot represent (the shapes listed in [Void columns without the table feature](#void-columns-without-the-table-feature)):
- a `void` directly inside an `array` or `map` (at any nesting level): the `void` element or value is structural and must be stored as `UNKNOWN`.

> - a `struct` whose fields are all `void`, or a table whose columns are all `void`: at least one of the `void` columns is structural and must be stored as `UNKNOWN`.

Reviewer 3: Shall we explicitly state that the choice of "which one/ones" is up to the writer?

Author: Updated.

A `void` column in any other position is never structural: it can be omitted, and does not require the feature. A schema that contains one of the shapes above is said to **require** the `voidType` feature.

> ### Writer Requirements for Void Type

Reviewer 3: Shall we also explain how writers should handle statistics for VOID columns?

Author: I don't think we need any special handling for stats, so the default rules from the protocol should be sufficient.

When Void Type is supported (when the `writerFeatures` field of a table's `protocol` action contains `voidType`), writers:
- must store the table's structural `void` columns as `UNKNOWN` (see [Structural void columns](#structural-void-columns)).

> - may store any non-structural `void` column either by omission or as an `UNKNOWN` column.

Reviewer 3: Why don't we force writers to always write non-structural `void` columns by omission?

Author: Because that would make the protocol more difficult to implement and is not strictly necessary. It only makes dropping the feature easier if we have less UNKNOWNs.

But I'll update it to "should" to show it's the preferred behavior.

### Reader Requirements for Void Type

When Void Type is supported (when the `readerFeatures` field of a table's `protocol` action contains `voidType`), readers:

> - must recognize and tolerate a `void` data type anywhere in a Delta table schema.

Reviewer 2: This phrasing is a bit weird. "must allow"

> - must read a `void` column stored as `UNKNOWN` as an all-`NULL` column.

Reviewer 2: specify Parquet here, since this seems targeted. 

Although maybe a more neutral way to frame this is "must return only null values for columns defined as void in the table schema". Whether it's omitted or materialized, the actual behaviour is that. It's less about the underlying data files' schema than it is about the actual Delta table's schema.

Author: Reworded it to not mention any type and just say return all null independent of representation.

##### Void columns without the table feature

When the `materializedVoidType` feature is not supported, `void` columns can only be omitted. Readers must reconstruct every omitted `void` column as an all-`NULL` column. Writers must reject operations that would write new data files when the table schema contains any of the following shapes, in which omitting the `void` column(s) would leave nowhere to record the nullability or length of an enclosing value:
- a `void` type used directly as an `array` element or `map` key or value, at any nesting level;
- a `struct` (at any nesting level) whose fields are all `void`; or
- a table whose non-partition columns are all `void`.

These restrictions are stated in terms of the **table schema**, not the schema of any individual data file. A table with a restricted schema can still be created, altered through metadata-only operations, and read. It can be made writable by evolving its schema - for example, by changing a `void` column to another type - or by enabling the `materializedVoidType` feature.

##### Void Type table feature

> The `materializedVoidType` table feature lifts the restrictions above by allowing the affected schema shapes. The `void` columns that cannot be omitted - the **structural** `void` columns - are instead stored using the Parquet `UNKNOWN` logical type. A `void` partition column is never structural and should be omitted.

Reviewer: > The `void` columns that cannot be omitted - the **structural** `void` columns - are instead stored using the Parquet `UNKNOWN` logical type.

This sentence should be split in two: 1. Only **structural** `void` columns SHOULD be materialized. 2. Materialized `void` columns MUST use the Parquet `UNKNOWN` logical type.

1. A client that supports `materializedVoidType` is guaranteed to correctly read and write `void` columns that rely on the missing columns mechanism. Enabling the feature for a table that only uses such columns is **optional**; a user may choose to enable it so that only clients capable of handling `void` columns correctly interact with the table.
2. A client that supports `materializedVoidType` is also guaranteed to correctly read and write structural `void` columns stored as `UNKNOWN`. Enabling the feature is **required** for any operation that would write new data files when the table schema requires the `UNKNOWN` representation. A writer may add the feature in the same commit that first introduces such data files, but must not commit an `UNKNOWN` column unless the feature is supported by the resulting table version.

###### Structural void columns

A `void` column is **structural** when it cannot be omitted because omitting it would leave an enclosing `struct`, `array`, or `map` (or the table) with nothing written to the data file. This arises only in the schema shapes listed in [Void columns without the table feature](#void-columns-without-the-table-feature):

> - a `void` used directly as an `array` element or `map` key or value (at any nesting level) is structural and must be stored as `UNKNOWN`;

Reviewer: Nit: The must be stored as `UNKNOWN` part is repeated three times. Let's split it out.

- for a `struct` whose fields are all `void`, the writer must choose at least one of its `void` fields as structural and store it as `UNKNOWN`; and

> - for a table whose non-partition columns are all `void`, the writer must choose at least one non-partition `void` column as structural and store it as `UNKNOWN`.

Reviewer: What if non-void partition columns are materialized in the data files?

Author: Technically in that case, we don't need to have any structural void columns, but phrasing that complicates the Protocol and the implementations. I thought it would be simpler to always enforce this rule independent of the partitioning columns and whether we materialize the partitioning columns.

A `void` column in any other position is non-structural: it can be omitted and does not require the feature. A schema that contains one of the shapes above is said to **require** the `materializedVoidType` feature.

###### Writer Requirements for Void Type

> When Void Type is supported (when the `writerFeatures` field of a table's `protocol` action contains `materializedVoidType`), writers:

Reviewer: ```suggestion
When materialized Void Type is supported (when the `writerFeatures` field of a table's `protocol` action contains `materializedVoidType`), writers:
```

###### Interaction with Materialize Partition Columns

The `materializePartitionColumns` writer feature requires writers to materialize only non-`void` partition columns. Writers should omit `void` partition columns whether or not `materializedVoidType` is also supported. A `void` partition column is not structural, and this exception does not relax the restrictions in [Void columns without the table feature](#void-columns-without-the-table-feature) or the structural-column requirements above.

###### Reader Requirements for Void Type

When Void Type is supported (when the `readerFeatures` field of a table's `protocol` action contains `materializedVoidType`), readers:
- must allow a `void` data type anywhere in a Delta table schema;
- must return only `NULL` values for a column defined as `void` in the table schema, regardless of whether it is omitted or materialized; and
- must, within a single scan, correctly combine data files that represent the same column differently - omitted, written as `UNKNOWN`, or (after a type change) written with a concrete type - into the requested read schema.

###### Removing the Void Type feature

> Because clients that do not support the feature cannot read `UNKNOWN` columns, removing `materializedVoidType` requires that the table no longer depend on the `UNKNOWN` representation. In the version that removes `materializedVoidType` from the `writerFeatures` and `readerFeatures` fields of the table's `protocol` action, writers:

Reviewer: Nit: "Because clients that do not support the feature **are not guaranteed to be able to** read `UNKNOWN` columns

- must ensure that the table schema does not require the feature - it must contain none of the shapes in [Structural void columns](#structural-void-columns); and

> - must ensure that no data file reachable by the table (including via time travel within the retained history) contains a column stored as `UNKNOWN`. This may require rewriting existing data files and/or truncating the table history so that every `void` column is represented by omission.

Reviewer: This needs some debate. I suspect that most readers today would be able to correctly handle void columns stored as `UNKNOWN` and if not they are highly likely to error out. Requiring a full rewrite of the table (instead of just dropping the column, which is a metadata rewrite) adds significant cost to the downgrade path.

Author: It's hard to say what other readers would do because UNKNOWN is a logical type that is put on some other physical type (like integer or boolean). Before Spark Parquet reader supported UNKNOWN, it used to ignore the logical type annotation and instead inferred type based on the physical type (so read the column as integer or boolean). Delta Spark connector would probably have failed in this scenario, as Delta would request a Void column but Parquet reader would read a different type causing mismatch.

UNKNOWN is a bit of an obscure type and Parquet readers have been slow to adopt it (at least the ones I know of), it's questionable whether most readers can handle it or not. But if we assume that the readers error out when they can't handle UNKNOWN properly, would it be acceptable? The error message will be something obscure and I can imagine it can confuse users.

> # Void Type

Reviewer 4: Materialized Void type? both title and name of file here and in readme?

Author: Thanks, missed it before. Changed it in relevant places.

- a `void` used directly as an `array` element or `map` key or value (at any nesting level) is structural;
- for a `struct` whose fields are all `void`, the writer must select one or more of its `void` fields as structural; and
- for a table whose non-partition columns are all `void`, the writer must select one or more non-partition `void` columns as structural.

A `void` column in any other position is non-structural and must be omitted. In particular, a `void` partition column is never structural. A schema that contains one of the shapes above is said to **require** the `materializedVoidType` feature.

###### Writer Requirements for Void Type

When materialized Void Type is supported (when the `writerFeatures` field of a table's `protocol` action contains `materializedVoidType`), writers:
- must write the table's structural `void` columns to data files (see [Structural void columns](#structural-void-columns)); and
- must omit every non-structural `void` column from data files, regardless of any other supported table feature or table property.

###### Interaction with Materialize Partition Columns

> The `materializePartitionColumns` writer feature requires writers to materialize only non-`void` partition columns. Writers must omit `void` partition columns whether or not `materializedVoidType` is also supported. A `void` partition column is not structural, and this exception does not relax the restrictions in [Void columns without the table feature](#void-columns-without-the-table-feature) or the structural-column requirements above.

Reviewer 5: i see this mentioned in too many places. Should we deduplicate some of this?

Author: I reduced the number of times we repeat information.

###### Reader Requirements for Void Type

When Void Type is supported (when the `readerFeatures` field of a table's `protocol` action contains `materializedVoidType`), readers:
- must allow a `void` data type anywhere in a Delta table schema;
- must return only `NULL` values for a column defined as `void` in the table schema, regardless of whether it is omitted or materialized; and

> - must, within a single scan, correctly combine data files that represent the same column differently - omitted, written as `UNKNOWN`, or (after a type change) written with a concrete type - into the requested read schema.

Reviewer 5: omitted or written => phrasing. can this AI phrasing be passed through a humanize skill
might help reduce the verbosity

Author: > omitted or written => phrasing

I'm not sure what you mean by this. Can you please clarify?

> can this AI phrasing be passed through a humanize skill might help reduce the verbosity

I ran humanize skill, it didn't help with duplication/repetition. In general, I think it made it worse and more AI-like. I kept some of the changes it made, but not everything.

###### Removing the Void Type feature

Because clients that do not support the feature are not guaranteed to be able to read `UNKNOWN` columns that correspond to columns in the current table schema, removing `materializedVoidType` requires that the current table version no longer depend on the `UNKNOWN` representation. In the version that removes `materializedVoidType` from the `writerFeatures` and `readerFeatures` fields of the table's `protocol` action, writers:
- must ensure that the table schema does not require the feature - it must contain none of the shapes in [Structural void columns](#structural-void-columns); and
- must ensure that every `UNKNOWN` column in a data file read by the current table version either corresponds to a column that has been dropped from the current table schema or is removed by rewriting the data file. Changing a `void` column to a concrete type does not satisfy this requirement by itself; data files that store that column as `UNKNOWN` must be rewritten.

> An `UNKNOWN` column that corresponds to a column dropped from the current table schema may remain in a data file. Readers of the current table version do not request the dropped column, while historical table versions that expose the column continue to advertise `materializedVoidType`. Therefore, the feature may be removed without rewriting data files when every column represented as `UNKNOWN` has been dropped from the current table schema.

Reviewer 5: this is an optimization suggestion right? Does the protocol generally capture this? Or the implementation
is left to the connectors?

Author: Yes, the protocol doesn't really mention this as a general rule. There are certain sentences that require some columns to be ignored, like CDF `_change_type` for readers that don't understand it. I'm changing the phrasing here to make it similar to that CDF rule.
